crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/json/CacheMetadataTemplate.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.cache.json;
+
+import com.google.cloud.tools.crepecake.json.JsonTemplate;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * JSON template for storing metadata about the cache.
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * {
+ *   "layers": [
+ *     {
+ *       "type": "BASE",
+ *       "reference": {
+ *         "size": 631,
+ *         "digest": "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef",
+ *         "diffId": "sha256:b56ae66c29370df48e7377c8f9baa744a3958058a766793f821dadcb144a4647"
+ *       },
+ *       "existsOn": ["gcr.io/my-project/my-image"]
+ *     },
+ *     ...
+ *     {
+ *       "type": "CLASSES",
+ *       "reference": {
+ *         "size": 223,
+ *         "digest": "sha256:8c662931926fa990b41da3c9f42663a537ccd498130030f9149173a0493832ad",
+ *         "diffId": "sha256:a3f3e99c29370df48e7377c8f9baa744a3958058a766793f821dadcb144a8372"
+ *       },
+ *       "existsOn": ["gcr.io/my-project/my-image"],
```

Comment:
existsOn sounds like it should be a runtime check?

Reply:
This field keeps track of which image tags this layer has been pushed to, and will be used as an indicator as to whether or not a layer should be pushed. Usually, layers should not be deleted from a registry, and this should directly reflect the existence of the layer on the registry. If the layer has been deleted, however, this edge case can be detected when the final manifest is pushed (at which point it will error and we can clear the cache and rebuild).

Perhaps I should rename this `pushedTo`?

Reply:
How expensive is it to check if a layer exists remotely though? My only concern is that this optimization doesn't seem completely safe.

Reply:
It's just extra HEAD requests. It should be safe since a registry won't accept a manifest with nonexistent layers.

Reply:
Oh, wait I misread that, are we going to handle it correctly for users? What do you mean by error?

Reply:
When the registry receives a manifest with nonexistent layers, it will send back an error response.

When that happens, we can either:
- fail and tell the user to clear their cache and retry, or
- push all the layers and the manifest again